File: heaps/1851_hard/min_intervals.cpp
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

vector<int> findMinIntervals(vector<vector<int>> &intervals, vector<int> &queries)
{ // LeetCode Q.1851.
    // Min heap. Format: {interval start, size}.
    priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> startsSizesHeap;

    // Min heap. Format: {size, interval end}.
    priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> sizesEndsHeap;

    for (auto interval : intervals)
        startsSizesHeap.push({interval[0], interval[1] + 1 - interval[0]});

    vector<pair<int, int>> queriesIndices; // Format: {query num, query idx}.

    for (int queryIdx = 0; queryIdx < queries.size(); queryIdx++)
        queriesIndices.push_back({queries[queryIdx], queryIdx});

    sort( // Sort by ascending query nums.
        queriesIndices.begin(), queriesIndices.end(),
        [](const pair<int, int> &a, const pair<int, int> &b)
        { return a.first < b.first; });

    vector<int> answers(queries.size(), -1); // Default to -1.

    for (auto [queryNum, queryIdx] : queriesIndices)
    {
        while (!startsSizesHeap.empty() && startsSizesHeap.top()[0] <= queryNum)
        {
            int start = startsSizesHeap.top()[0], size = startsSizesHeap.top()[1];
            sizesEndsHeap.push({size, start + size - 1}); // Format: {size, interval end}.
            startsSizesHeap.pop();
        }

        while (!sizesEndsHeap.empty() && sizesEndsHeap.top()[1] < queryNum)
            sizesEndsHeap.pop();

        if (!sizesEndsHeap.empty())
            answers[queryIdx] = sizesEndsHeap.top()[0];
    }

    return answers;
}

File: heaps/1851_hard/min_intervals.py
import heapq


def find_min_intervals(intervals: list[list[int]], queries: list[int]) -> list[int]:  # LeetCode Q.1851.
    starts_sizes_heap = []  # Min heap. Format: (interval start, size).
    sizes_ends_heap = []  # Min heap. Format: (size, interval end).

    for start, end in intervals:
        heapq.heappush(starts_sizes_heap, (start, end + 1 - start))

    queries_indices = []  # Format: (query num, query idx).
    for query_idx, query_num in enumerate(queries):
        queries_indices.append((query_num, query_idx))

    queries_indices.sort()  # Sort by ascending query nums.

    answers = [-1] * len(queries)  # Default to -1.
    for query_num, query_idx in queries_indices:
        while starts_sizes_heap and starts_sizes_heap[0][0] <= query_num:
            start, size = starts_sizes_heap[0]
            heapq.heappush(sizes_ends_heap, (size, start + size - 1))  # Format: (size, interval end).
            heapq.heappop(starts_sizes_heap)

        while sizes_ends_heap and sizes_ends_heap[0][1] < query_num:
            heapq.heappop(sizes_ends_heap)

        if sizes_ends_heap:
            answers[query_idx] = sizes_ends_heap[0][0]

    return answers
